SSH communicator panics when HTTP proxy CONNECT response fails (nil res.Body.Close)

### Community note

> [!TIP]
> 👋 Hi there, OpenTofu community! The OpenTofu team prioritizes issues based on upvotes. Please make sure to upvote this issue and describe how it affects you in detail in the comments to show your support.

### OpenTofu Version

```shell
OpenTofu v1.12.4
on darwin_arm64

Also confirmed present on main @ 3cc4b302082885582662d6287bc2ecf8d8bb91d7
(internal/communicator/ssh/http_proxy.go)
```

### OpenTofu Configuration Files

Any `connection` block that sets an HTTP proxy for SSH (used by `remote-exec` / `file` provisioners):

```hcl
resource "null_resource" "proxy_ssh" {
  connection {
    type     = "ssh"
    host     = "10.0.0.5"
    user     = "ubuntu"
    private_key = file("~/.ssh/id_rsa")

    proxy_scheme = "http"
    proxy_host   = "127.0.0.1"
    proxy_port   = 18080
  }

  provisioner "remote-exec" {
    inline = ["true"]
  }
}
```

`proxy_host` must point at something that accepts TCP but does not return a valid HTTP CONNECT response (closed connection, non-HTTP peer, truncated reply, etc.).

### Debug Output

No full `TF_LOG=trace` gist: OpenTofu panics during the SSH dial through the proxy, so apply does not finish with a normal diagnostic stream.

Minimal standalone proof of the same library call pattern used in `http_proxy.go` (TCP peer accepts then resets; `http.ReadResponse` returns `res == nil` and a non-nil error; then `res.Body.Close()` panics):

```text
ReadResponse: res=<nil> err=read tcp 127.0.0.1:...: connection reset by peer
About to call res.Body.Close() on error path...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x40 ...]
```

### Expected Behavior

When the HTTP proxy fails to complete CONNECT (EOF, reset, malformed response), OpenTofu should return a normal dial/proxy error to the provisioner, for example something like:

```text
Error: ... failed to dial ...: <underlying ReadResponse or proxy error>
```

The process should not crash.

### Actual Behavior

On the `http.ReadResponse` error path, OpenTofu calls `res.Body.Close()` while `res` is often nil, which panics:

```text
panic: runtime error: invalid memory address or nil pointer dereference
```

Relevant code on main:

https://github.com/opentofu/opentofu/blob/3cc4b302082885582662d6287bc2ecf8d8bb91d7/internal/communicator/ssh/http_proxy.go#L102-L107

```go
res, err := http.ReadResponse(bufio.NewReader(c), req)

if err != nil {
        res.Body.Close()  // res is often nil here
        c.Close()
        return nil, err
}
```

Go's `http.ReadResponse` can return `(nil, err)` when the response cannot be parsed at all. Closing the body without a nil check turns a recoverable proxy failure into a process-wide crash.

The successful path already closes the body correctly a few lines below; only the error path is unsafe. The TCP connection `c` is still closed on that path (good); the bug is solely the unconditional `res.Body.Close()`.

### Steps to Reproduce

1. Start a TCP listener that accepts a connection and immediately closes it (or resets), without speaking HTTP. Example:

```bash
# one-shot "bad proxy": accept once, then close
python3 -c '
import socket
s=socket.socket(); s.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR,1)
s.bind(("127.0.0.1",18080)); s.listen(1)
c,_=s.accept(); c.close(); s.close()
'
```

2. Use the configuration above (or any provisioner `connection` with `proxy_host` / `proxy_port` pointing at that listener).
3. Run `tofu apply` (or any command that runs the provisioner and dials SSH via that HTTP proxy).
4. Observe a nil pointer panic instead of a dial error.

Alternatively, any real HTTP proxy that drops the CONNECT handshake mid-stream triggers the same path.

### Additional Context

- Triggered only when `connection.proxy_host` is set for the SSH communicator (`internal/communicator/ssh/provisioner.go` builds a `proxyInfo` and dials through `newHttpProxyConn` / `proxyDialer.Dial`).
- Affects provisioners that use SSH with an explicit HTTP(S) proxy (`proxy_scheme` defaults to `http` when empty).
- Suggested fix (conceptual only; not a PR): on the `ReadResponse` error path, close the body only when `res != nil && res.Body != nil`, and always close `c`. Example shape:

```go
res, err := http.ReadResponse(bufio.NewReader(c), req)
if err != nil {
        if res != nil && res.Body != nil {
                res.Body.Close()
        }
        c.Close()
        return nil, err
}
```

### References

- Code: [`internal/communicator/ssh/http_proxy.go`](https://github.com/opentofu/opentofu/blob/3cc4b302082885582662d6287bc2ecf8d8bb91d7/internal/communicator/ssh/http_proxy.go#L102-L107) (`proxyDialer.Dial`, CONNECT response handling)
- Connection wiring: `internal/communicator/ssh/provisioner.go` (`prepareSSHConfig` when `ProxyHost != ""`)
- No existing open issue found for this specific nil panic on the CONNECT error path

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -55,6 +55,7 @@ BUG FIXES:
 - `tofu plan -out` no longer fails when the plan includes a resource with `lifecycle { destroy = false }` that needs replacement, which previously errored with `invalid change action ForgetThenCreate`. ([#4324](https://github.com/opentofu/opentofu/issues/4324))
 - `connection.script_path` is escaped correctly not allowing anymore additional commands to be executed on the remote host together with the script path indicated by the argument. ([#4330](https://github.com/opentofu/opentofu/pull/4330))
 - `tofu plan`: Fixed Incorrect warnings produced for `tofu plan -replace=ADDR`. ([#4368](https://github.com/opentofu/opentofu/issues/4368))
+- SSH connections through an HTTP proxy now report a connection error instead of crashing when the proxy fails to answer the CONNECT request. ([#4358](https://github.com/opentofu/opentofu/issues/4358))
 
 ## Previous Releases
 

--- a/internal/communicator/ssh/http_proxy.go
+++ b/internal/communicator/ssh/http_proxy.go
@@ -102,7 +102,6 @@ func (p *proxyDialer) Dial(network, addr string) (net.Conn, error) {
 	res, err := http.ReadResponse(bufio.NewReader(c), req)
 
 	if err != nil {
-		res.Body.Close()
 		c.Close()
 		return nil, err
 	}

--- a/internal/communicator/ssh/http_proxy_test.go
+++ b/internal/communicator/ssh/http_proxy_test.go
@@ -1,0 +1,105 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package ssh
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+	"strings"
+	"testing"
+
+	"golang.org/x/net/proxy"
+)
+
+func TestProxyDialer_Dial_connectResponseError(t *testing.T) {
+	// A proxy that accepts the CONNECT request and then closes without
+	// answering must surface a dial error, not crash the process
+	addr := fakeProxy(t, func(net.Conn) {})
+
+	pd := &proxyDialer{
+		proxy:   *newProxyInfo(addr, "http", "", ""),
+		forward: proxy.Direct,
+	}
+
+	conn, err := pd.Dial("tcp", "192.0.2.1:22")
+	if err == nil {
+		conn.Close()
+		t.Fatal("expected an error, got nil")
+	}
+}
+
+func TestProxyDialer_Dial_success(t *testing.T) {
+	// A proxy that answers the CONNECT request with 200 must yield a connection
+	addr := fakeProxy(t, func(c net.Conn) {
+		_, _ = c.Write([]byte("HTTP/1.1 200 Connection established\r\n\r\n"))
+	})
+
+	pd := &proxyDialer{
+		proxy:   *newProxyInfo(addr, "http", "", ""),
+		forward: proxy.Direct,
+	}
+
+	conn, err := pd.Dial("tcp", "192.0.2.1:22")
+	if err != nil {
+		t.Fatalf("Dial returned error: %v, want nil", err)
+	}
+	defer conn.Close()
+}
+
+func TestProxyDialer_Dial_nonOKStatus(t *testing.T) {
+	// A proxy that rejects the CONNECT request must surface its status code as a dial error
+	addr := fakeProxy(t, func(c net.Conn) {
+		_, _ = c.Write([]byte("HTTP/1.1 407 Proxy Authentication Required\r\nContent-Length: 0\r\n\r\n"))
+	})
+
+	pd := &proxyDialer{
+		proxy:   *newProxyInfo(addr, "http", "", ""),
+		forward: proxy.Direct,
+	}
+
+	conn, err := pd.Dial("tcp", "192.0.2.1:22")
+	if err == nil {
+		conn.Close()
+		t.Fatal("expected an error, got nil")
+	}
+	if !strings.Contains(err.Error(), "407") {
+		t.Fatalf("error = %v, want it to mention 407", err)
+	}
+}
+
+// fakeProxy starts an HTTP proxy on the loopback interface and returns its address.
+// It accepts one connection, reads the CONNECT request, calls respond and then
+// closes: a respond that writes nothing models a proxy that hangs up without answering
+func fakeProxy(t *testing.T, respond func(net.Conn)) string {
+	t.Helper()
+
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		conn, err := ln.Accept()
+		if err != nil {
+			return
+		}
+		defer conn.Close()
+
+		_, _ = http.ReadRequest(bufio.NewReader(conn))
+		respond(conn)
+	}()
+
+	t.Cleanup(func() {
+		ln.Close()
+		<-done
+	})
+
+	return ln.Addr().String()
+}

